crates/audio/src/spatial.rs

```diff
@@ -0,0 +1,292 @@
+) -> (f64, f64) {
+    let cam_right = camera.right();
+    let sound_dir = (sound_position - camera.translation()).normalize();
+    let pan = cam_right.dot(sound_dir) * 0.5 + 0.5;
+
+    let attenuation_factor = match attenuation.attenuation {
+        Some(Attenuation::LinearFromFocus) => {
+            let distance_from_focus = focus.point().distance(sound_position);
+            let biased_zoom_distance = focus.distance().inner() + SOUND_LINEAR_FALLOFF_BIAS;
+
+            let distance_ratio = distance_from_focus / biased_zoom_distance;
+            // Clamped linear function passing through (start, 1) and (end, 0) in X units of camera zoom distance.
+            ((SOUND_LINEAR_FALLOFF_END_RATIO - distance_ratio) / SOUND_LINEAR_FALLOFF_LENGTH)
+                .clamp(0.0, 1.0)
+        }
+        Some(Attenuation::InverseSquare) => {
+            let distance_from_camera_squared =
+                camera.translation().distance_squared(sound_position);
+            // Anything closer than 70% of zoom distance is at full volume.
+            let min_distance_squared = (0.7 * focus.distance().inner()).powi(2);
+
+            (min_distance_squared / distance_from_camera_squared).min(1.)
+        }
+        Some(Attenuation::InverseSquareFromFocus) => {
+            let distance_from_focus_squared = focus.point().distance_squared(sound_position);
+            // Anything closer than 70% of zoom distance is at full volume.
+            let min_distance_squared = (0.7 * focus.distance().inner()).powi(2);
+
+            (min_distance_squared / distance_from_focus_squared).min(1.)
+        }
+        None => 1.,
+    };
+
+    let occlusion_factor = match attenuation.occlusion {
+        Some(Occlusion::HardFrustum) => {
+            if let Some(ndc) = camera_config.world_to_ndc(camera, sound_position) {
+                if (-1.0..=1.0).contains(&ndc.x)
+                    && (-1.0..=1.0).contains(&ndc.y)
+                    && (0.0..=1.0).contains(&ndc.z)
+                {
+                    1.
+                } else {
+                    0.
+                }
+            } else {
+                0.
+            }
+        }
+        Some(Occlusion::Angle) => {
+            // Volume is 1 from 0-25.5 degrees from the view direction, then linearly goes to 0 from 25.5-45 degrees.
+            // Angle is calculated from a point slightly behind the camera.
+            let camera_offset = camera.translation() + camera.back() * SOUND_VIEW_ANGLE_OFFSET;
+            let angle = (sound_position - camera_offset).angle_between(camera.forward());
+
+            // Slope parameter of linear function (-1 unit in 25.5 degrees or pi/8)
+            const SLOPE: f32 = -1.0 / (PI / 8.);
+            // Constant parameter of linear function
+            const CONSTANT: f32 = 2.0;
+            (SLOPE * angle + CONSTANT).clamp(0.0, 1.0)
+        }
+        None => 1.,
+    };
+
+    let volume = attenuation_factor * occlusion_factor;
```

Comment:
I would clamp this instead of the operands individually so that something e.g. very close but at a bad angle doesn't become unnecessarily silent.